front/pages/api/w/[wId]/vaults/[vId]/data_sources/index.ts

```diff
@@ -0,0 +1,633 @@
+  DataSourceViewType,
+  PlanType,
+  UserType,
+  WithAPIErrorResponse,
+  WorkspaceType,
+} from "@dust-tt/types";
+import {
+  CONNECTOR_PROVIDERS,
+  ConnectorConfigurationTypeSchema,
+  ConnectorsAPI,
+  CoreAPI,
+  DEFAULT_EMBEDDING_PROVIDER_ID,
+  DEFAULT_QDRANT_CLUSTER,
+  dustManagedCredentials,
+  EMBEDDING_CONFIGS,
+  ioTsParsePayload,
+  isDataSourceNameValid,
+  sendUserOperationMessage,
+  WebCrawlerConfigurationTypeSchema,
+} from "@dust-tt/types";
+import { isLeft } from "fp-ts/lib/Either";
+import * as t from "io-ts";
+import * as reporter from "io-ts-reporters";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import config from "@app/lib/api/config";
+import { getDataSource } from "@app/lib/api/data_sources";
+import { withSessionAuthenticationForWorkspace } from "@app/lib/api/wrappers";
+import type { Authenticator } from "@app/lib/auth";
+import { getOrCreateSystemApiKey } from "@app/lib/auth";
+import {
+  getDefaultDataSourceDescription,
+  getDefaultDataSourceName,
+  isConnectionIdRequiredForProvider,
+  isConnectorProviderAllowedForPlan,
+  isConnectorProviderAssistantDefaultSelected,
+  isValidConnectorSuffix,
+} from "@app/lib/connector_providers";
+import { DataSourceResource } from "@app/lib/resources/data_source_resource";
+import { DataSourceViewResource } from "@app/lib/resources/data_source_view_resource";
+import { VaultResource } from "@app/lib/resources/vault_resource";
+import { ServerSideTracking } from "@app/lib/tracking/server";
+import { isDisposableEmailDomain } from "@app/lib/utils/disposable_email_domains";
+import logger from "@app/logger/logger";
+import { apiError } from "@app/logger/withlogging";
+
+export const PostDataSourceWithProviderRequestBodySchema = t.type({
+  provider: getConnectorProviderCodec(),
+  name: t.union([t.string, t.undefined]),
+  connectionId: t.union([t.string, t.undefined]),
+  configuration: ConnectorConfigurationTypeSchema,
+});
+// Sorcery: Create a union type with at least two elements to satisfy t.union
+export function getConnectorProviderCodec(): t.Mixed {
+  const [first, second, ...rest] = CONNECTOR_PROVIDERS;
+  return t.union([
+    t.literal(first),
+    t.literal(second),
+    ...rest.map((value) => t.literal(value)),
+  ]);
+}
+
+const PostDataSourceWithoutProviderRequestBodySchema = t.type({
+  name: t.string,
+  description: t.string,
+});
+
+const PostDataSourceRequestBodySchema = t.union([
+  PostDataSourceWithoutProviderRequestBodySchema,
+  PostDataSourceWithProviderRequestBodySchema,
+]);
+
+export type PostVaultDataSourceResponseBody = {
+  dataSource: DataSourceType;
+  dataSourceView: DataSourceViewType;
+};
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<WithAPIErrorResponse<PostVaultDataSourceResponseBody>>,
+  auth: Authenticator
+): Promise<void> {
+  const owner = auth.workspace();
+  const plan = auth.plan();
+  const user = auth.user();
+
+  if (!owner || !plan || !user || !auth.isUser()) {
```

Comment:
This condition is not required and already checked by `withSessionAuthenticationForWorkspace`